### PR TITLE
Use URIs for devices and files

### PR DESCRIPTION
**Pull request by gavv**
This PR switches roc_sndio and tools from using driver name plus input/output name to using a single URI, both for devices and files. The concept of drivers becomes completely hidden from user.

See #258.

### CODE REVIEW

File: src/modules/roc_sndio/ibackend.h [hunk 1]
```diff
@@ -59,7 +58,7 @@ class IBackend {
                                  const Config& config) = 0;
 
     //! Append supported dirvers to the list.
```
Comment (dshil) [inline src/modules/roc_sndio/ibackend.h]: s/dirvers/drivers

File: docs/sphinx/manuals/roc_recv.rst [hunk 2]
```diff
@@ -138,31 +148,37 @@ Output to the default ALSA device:
...
 
-    $ roc-recv -vv -s rtp+rs8m::10001 -r rs8m::10002 -o ./file.wav
+    $ roc-recv -vv -o file:./output.wav -s rtp+rs8m::10001 -r rs8m::10002
 
 Output to stdout in WAV format:
 
 .. code::
 
-    $ roc-recv -vv -s rtp+rs8m::10001 -r rs8m::10002 -d wav -o - > ./file.wav
+    $ roc-recv -vv -o file:- -f wav -s rtp+rs8m::10001 -r rs8m::10002 > ./output.wav
```
Comment (dshil) [inline docs/sphinx/manuals/roc_recv.rst]: Could you also add an example for a non-special file without an extension:

```bash
roc-recv -vv -o file:./file -f wav -s rtp+rs8m::10001 -r rs8m::10002
```

File: src/tests/roc_core/test_string_list.cpp [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+TEST(string_list, nextof) {
+    StringList sl(allocator);
+
+    CHECK(sl.push_back("foo"));
+    CHECK(sl.push_back("bar"));
+    CHECK(sl.push_back("baz"));
+
+    LONGS_EQUAL(3, sl.size());
+
+    const char* str;
```
Comment (dshil) [inline src/tests/roc_core/test_string_list.cpp]:
```cpp
    const char* str = NULL;
```

?

File: src/tests/roc_core/test_string_list.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        if (prev_front != sl.front()) {
+            num_reallocs++;
+            prev_front = sl.front();
+        }
+
+        LONGS_EQUAL(n + 1, sl.size());
+        LONGS_EQUAL(expected_reallocs[n], num_reallocs);
+    }
+}
+
```
Comment (dshil) [inline src/tests/roc_core/test_string_list.cpp]: Lets add some tests for empty strings, something similar to:

```cpp
TEST(string_list, empty_strings) {
    StringList sl(allocator);

    CHECK(sl.push_back(""));
    CHECK(sl.push_back(""));
    CHECK(sl.push_back(""));

    LONGS_EQUAL(3, sl.size());

    const char* str = NULL;

    str = sl.front();
    STRCMP_EQUAL("", str);

    str = sl.nextof(str);
    STRCMP_EQUAL("", str);

    str = sl.nextof(str);
    STRCMP_EQUAL("", str);

    str = sl.nextof(str);
    CHECK(str == NULL);
}
```

?

Comment (dshil) [reply src/tests/roc_core/test_string_list.cpp]: Probably we can also add the following test:

```cpp
TEST(string_list, push_back_twice) {
    StringList sl(allocator);

    const char* str = "foo";

    sl.push_back(str);
    sl.push_back(str);

    LONGS_EQUAL(2, sl.size());

    const char* next = sl.front();
    STRCMP_EQUAL("foo", next);

    next = sl.nextof(next);
    STRCMP_EQUAL("foo", next);

    next = sl.nextof(next);
    CHECK(next == NULL);
}
```

to show that we do not touch the input string (for demonstration purpose).

File: docs/sphinx/manuals/roc_conv.rst [hunk 5]
```diff
@@ -14,18 +14,51 @@ Read audio stream from a file, transform it, and and write it to a another file.
...
---resampler-window=INT    Number of samples per resampler window
---poisoning               Enable uninitialized memory poisoning (default=off)
+-h, --help                  Print help and exit
+-V, --version               Print version and exit
+-v, --verbose               Increase verbosity level (may be used multiple times)
+-L, --list-supported        list supported schemes and formats
+-i, --input=FILE_URI        Input file URI
+-o, --output=FILE_URI       Output file URI
+--input-format=FORMAT       Force input file format
+-f, --output-format=FORMAT  Force output file format
```
[the remark this answers is not in the document]
Comment (gavv) [reply docs/sphinx/manuals/roc_conv.rst]: I couldn't think of a name (letter) for it that wouldn't be confusing. Ideas?

File: src/tests/roc_core/test_string_list.cpp (hunk 4 above)
Comment (gavv) [reply src/tests/roc_core/test_string_list.cpp]: > to show that we do not touch the input string (for demonstration purpose).

What do you mean by "touching"?

Comment (dshil) [reply src/tests/roc_core/test_string_list.cpp]: I meant "touching" == "modifying".

File: docs/sphinx/manuals/roc_conv.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (dshil) [reply docs/sphinx/manuals/roc_conv.rst]: We can use only the full versions:

```
--input-format
--output-format
```

And remove `-f` support.

[the remark this answers is not in the document]
Comment (gavv) [reply docs/sphinx/manuals/roc_conv.rst]: OK. Done.

File: src/modules/roc_sndio/backend_dispatcher.h [hunk 6]
```diff
@@ -37,31 +38,35 @@ class BackendDispatcher : public core::NonCopyable<> {
...
 
 private:
     friend class core::Singleton<BackendDispatcher>;
 
     BackendDispatcher();
 
+    int select_driver_type_(const address::IoURI& uri) const;
+    const char* select_driver_name_(const address::IoURI& uri,
+                                    const char* force_format) const;
+    const char* select_inout_(const address::IoURI& uri) const;
```
[the remark this answers is not in the document]
Comment (gavv) [reply src/modules/roc_sndio/backend_dispatcher.h]: Done.

File: src/modules/roc_sndio/ibackend.h (hunk 1 above)
Comment (gavv) [reply src/modules/roc_sndio/ibackend.h]: Done.

File: src/tests/roc_core/test_string_list.cpp (hunk 3 above)
Comment (gavv) [reply src/tests/roc_core/test_string_list.cpp]: Fixed.

File: src/tests/roc_core/test_string_list.cpp (hunk 4 above)
Comment (gavv) [reply src/tests/roc_core/test_string_list.cpp]: Done.

File: docs/sphinx/manuals/roc_recv.rst (hunk 2 above)
Comment (gavv) [reply docs/sphinx/manuals/roc_recv.rst]: Done.

**Comment by gavv** [comment]
@tomd1990 FYI
